### PR TITLE
when checking for sidecar injection, check the label as well as the annotation

### PR DESCRIPTION
**Pull request by jmazzitelli**
fixes: https://github.com/kiali/kiali/issues/4268

UI PR: https://github.com/kiali/kiali-ui/pull/2230
Operator PR: https://github.com/kiali/kiali-operator/pull/413

### CODE REVIEW

**Comment by jmazzitelli** [comment]
I modified the code to only return a non-nil IstioInjectionAnnotation if the pod label is true or false OR if the pod annotation is false. If the pod annotation is true, that value is meaningless and so does not affect the value of IstioInjectionAnnotation.
Here's what it looks like if the label is true and if the annotation is true:

label == true
injection-true

annotation == true (meaningless value - you cannot opt in for sidecar injection with the annotation - it is opt-out only)
injection-false

**Comment by jmazzitelli** [comment]
I just realized something - the annotation has different behavior when running with Maistra. The annotation value of "true" DOES have meaning. I need to rethink that change I did.

**Comment by jmazzitelli** [comment]
this PR now provides a common util for use within the Kiali Server -- `status.IsMaistra()` -- this will return `true` if Kiali server is running within a Maistra environment.

This is used in this PR due to the different semantics of the auto-injection pod label and pod annotation between Maistra and non-Maistra environments. There are comments in the code to explain those different semantics so hopefully the code will make sense once you read the comments.

**Comment by jmazzitelli** [comment]
asking @lucasponce to review - this is adding a common API that can be used in the server anywhere. Just want to make sure you review it first before it gets merged.

**Comment by lucasponce** [comment]
@jmazzitelli I'm starting to review your PR, thanks, I guess you saw that you'd need also to update the UI too:
image

So, I think you need a combined PR for the frontend to merge all together (unless you want to merge them separated).

File: status/versions.go
Comment (lucasponce) [inline status/versions.go:46]: Minor var -> const if those are constants

File: models/workload.go
Comment (lucasponce) [inline models/workload.go:55]: Probably you may need to differentiate if you are using labels vs annotations and tell that to the UI which should build a patch depending of the case

https://github.com/kiali/kiali-ui/blob/master/src/components/IstioWizards/WizardActions.ts#L1407

Suggestion a new field with "IstioInjectionLabel" in the similar way, and let the UI to use one or another.

Comment (jmazzitelli) [reply models/workload.go:55]: I originally thought of that but decided against it because of:

1. the amount of change would be a lot more (it would also require a change to the operator also)
2. adding yet another config option that will likely never change and will likely never have a value different from the annotation label name

In other words, the name of this label is (and will always be) the same as the name of the annotation (`sidecar.istio.io/inject`). I don't see any benefit to creating a separate config option here because it will likely never be the case that the annotation label will be one name and the name of the label will be different than that.

Do you see a benefit to creating a separate one?

**Comment by jmazzitelli** [comment]
> So, I think you need a combined PR for the frontend to merge all together (unless you want to merge them separated).

The way the code is, the UI should not have to change. Because of the different semantics of the pod injection label between Maistra and upstream, that UI code can't use the label to add auto-injection. (see my comments above for a description of these differences).

And because the name of both the label and annotation are the same (and we don't have that extra superfluous config option you suggested earlier), the UI does not have to change anything when its displaying whether auto-injection is on or off for the workload.

If you can see a reason why we would need to add a second config option (istioInjectionLabel), then that would cause us to change the UI and operator (as well as this server PR)... this is why I decided not to add that extra config option - it isn't strictly needed and would only cause some churn in both operator and UI that is not needed as this PR is written now.

**Comment by lucasponce** [comment]
So, your suggestion is that the Workload action will always use the POd annotation in any case, even if the user decides to use the label in the workload instead.

I'm not initially against that, just thinking if there could be any clash.

**Comment by jmazzitelli** [comment]
> So, your suggestion is that the Workload action will always use the Pod annotation in any case, even if the user decides to use the label in the workload instead.

The only way we could have the UI use either label or annotation is if we were able to tell the UI that it is running in a Maistra or non-Maistra environment. This is because the pod label is completely meaningless in a Maistra environment (i.e. Maistra never looks at the label to know if it should inject the sidecar. Maistra ONLY ever looks at the annotatio). So it would not make sense (and in fact would probably be considered a bug) if the UI offered the user to add the label (as opposed to the annotation) when running in a Maistra environment.

Since upstream Istio also supports the annotation to enable injection, it seems the easiest thing to do is to leave the UI alone - the UI would be opinionated in this case. It will always use the annotation. Thus it will support both upstream Istio and Maistra as is.

Just to further explain the label and annotation support more clearly:


_UPSTREAM ISTIO_:
* **Namespace Label**: supports "true" and "false"
* **Pod Label**: supports "true" and "false"
* **Pod Annotation**: supports "false" only

_MAISTRA_:
* **Namespace Label**: not supported
* **Pod Label**: not supported
* **Pod Annotation**: "true" and "false"

**Comment by jmazzitelli** [comment]
OK, after discussing, I will change this and the UI to support different paths for non-Maistra and Maistra environs. If non-Maistra, the label will be used by the UI when it adds it to the workload; annotation will be used when on Maistra.

The UI will have an "isMaistra" flag for it to use here and anywhere else the UI needs to make this distinction.

**Review by jshaughn — COMMENTED** [review]
Visually looks OK to me.  I'll defer to @lucasponce for a final approval as he has been the main point of contact about the feature addition.

**Comment by jmazzitelli** [comment]
I added molecule testing for this - see the operator PR https://github.com/kiali/kiali-operator/pull/413

**Review by lucasponce — APPROVED** [review]
Reviewed and tested locally.

**Comment by lucasponce** [comment]
For QE: I think covering the upstream case would be enough (I think). Trying to build a Maistra version (Istio 1.9 based) with Kiali 1.40 can be more complex and probably this use case can be tested in the future in the context of SM 2.2.

**Pull request merged**
